astropy/io/fits/connect.py

```diff
@@ -272,30 +286,25 @@ def read_table_fits(
         # preserve null values.
         masked = mask = False
         fill_value = None
-        coltype = col.dtype.subdtype[0].type if col.dtype.subdtype else col.dtype.type
         if col.null is not None:
-            mask = data[col.name] == col.null
+            mask = arr == col.null
             # Return a MaskedColumn even if no elements are masked so
             # we roundtrip better.
             masked = True
             fill_value = col.null
         elif mask_invalid and issubclass(coltype, np.inexact):
-            mask = np.isnan(data[col.name])
+            mask = np.isnan(arr)
             fill_value = np.nan
         elif mask_invalid and issubclass(coltype, np.character):
-            mask = col.array == b""
+            mask = arr == b""
```

Comment:
Not sure if there was a good reason to use `col.array` instead of `data[col.name]` here. It returns a numpy array instead of a chararray but this doesn't matter to compute the mask I think.